[FLINK-16438][yarn] Make YarnResourceManager starts workers using WorkerResourceSpec requested by SlotManager

xintongsong opened the pull request:
## What is the purpose of the change

This PR is one step of FLINK-14106, making `YarnResourceManager` starts workers using `WorkerResourceSpec` requested by `SlotManager`.

This also means `YarnResourceManager` no longer:
- be aware of the default task executor resources
- assumes all workers are identical

This PR is based on #11323.

## Brief change log
- 5ee2a8ece3eefb4726618ed31a05a2e11737b57a..d7a0626255f2445cfe20dcb7bab1aede3af4b0d7: Commits from previous PR.
- 2f2053259d86240c56ad4498eadae237ab30e979..4e214f940b187704d0599b15380b78094b7ad7a8: Refactoring `YarnResourceManagerTest`, getting rid of using of `Mockito`.
- ef959ec03f964ceb0664388b4b72c8eac5773fbd: Introduce `WorkerSpecContainerResourceAdapter` for converting between Flink `WorkerResourceSpec` and Yarn container `Resource` in `YarnResourceManager`.
  - This is to address the problem that resource of containers allocated from Yarn can differs from what were requested.
- 785558f7c2944cb50e57e191a242c5e981ab42b7: `YarnResourceManager` starts workers with resources requested by `SlotManager`.
- 0c5eb8c3d46d2142f9dd64acfaba9d27102ad3f2: Remove unused `TaskExecutorProcessSpec` from `ActiveResourceManager`.

## Verifying this change

Added test cases in `YarnResourceManagerTest`, for:
- Validating behaviors of `WorkerSpecContainerResourceAdapter`.
- Validating `YarnResourceManager` dealing with workers with different resources.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

xintongsong commented:
@flinkbot run travis

KarmaGYZ commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -615,4 +632,85 @@ protected double getCpuCores(final Configuration configuration) {
+				normalize(taskExecutorProcessSpec.getCpuCores().getValue().intValue(), minVcore));
+			containerResourceToWorkerSpecs.computeIfAbsent(containerResource, ignored -> new ArrayList<>())
+				.add(workerResourceSpec);
+			containerMemoryToContainerResource.computeIfAbsent(containerResource.getMemory(), ignored -> new HashSet<>())
```
Seems we don't need to add it to `containerMemoryToContainerResource` if `matchVcores` is false.

xintongsong replied:
True, we don't need `containerMemoryToContainerResource` if not matching vcores. However, I would try avoid unnecessary `if-else` branches.

KarmaGYZ commented on flink-yarn/src/test/java/org/apache/flink/yarn/TestingYarnAMRMClientAsync.java:
```diff
@@ -0,0 +1,113 @@
+/**
+ * A Yarn {@link AMRMClientAsync} implementation for testing.
+ */
+public class TestingYarnAMRMClientAsync extends AMRMClientAsyncImpl<AMRMClient.ContainerRequest> {
```
Is there any benefit we could get from extending `AMRMClientAsyncImpl` instead of `AMRMClientAsync`? This class is annotated with `Unstable`.

xintongsong replied:
The problem is that `AMRMClientAsync` is an abstract class and we have to implement all its abstract methods if extending it. There are some abstract methods in later Hadoop versions that we cannot easily implement for early versions, because the absence of argument/return value type.

Although `AMRMClientAsyncImpl` is unstable, we are not really depending on its implementation. The methods we override are all declared in `AMRMClientAsync`, which is stable. So it should not be a problem.

xintongsong commented on flink-yarn/src/test/java/org/apache/flink/yarn/YarnResourceManagerTest.java:
```diff
@@ -584,6 +590,85 @@ public void testGetCpuExceedMaxInt() throws Exception {
 	}
 
+	@Test
+	public void testWorkerSpecContainerResourceAdapter_MatchVcores() {
```
I'm sure about this either. That's why I kept it, for discussions in the PR.

On one hand, I do find the following in the [code style guide](https://flink.apache.org/contributing/code-style-and-quality-formatting.html).
> Non-static fields/methods must be in lower camel case.

On the other hand, the underscores do provide better readability in the long test case name, and there are already test cases using them (e.g., MiniClusterConfigurationTest, CatalogTest, etc.).

KarmaGYZ replied:
I prefer to follow the code style guide if possible but I'm ok with it atm. Someone who has better English proficiency may give us another idea about the name.

xintongsong commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -615,4 +632,85 @@ protected double getCpuCores(final Configuration configuration) {
+		private final boolean matchVcores;
+		private final Map<WorkerResourceSpec, Resource> workerSpecToContainerResource;
+		private final Map<Resource, Collection<WorkerResourceSpec>> containerResourceToWorkerSpecs;
+		private final Map<Integer, Collection<Resource>> containerMemoryToContainerResource;
```
I think the upper bond of the amount of records really depends on how many different `WorkerResourceSpec` do we have.
If we want to clean the unused records up, the `WorkerSpecContainerResourceAdapter` will need `YarnResourceManager` to tell it which `WorkerResourceSpec` is no longer needed (all corresponding TMs are completed and no pending ones). ATM, I don't see the necessity for such comlexity.

xintongsong commented:
Rebased to resolve conflicts.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptionsInternal.java:
```diff
@@ -34,4 +37,24 @@
+	 * <p>If later we add another {@link SlotManager} implementation that may have TMs with different resources, we can
+	 * switch this option on only for the new SM, and the new SM can also be available on Hadoop 2.6+ only.
+	 */
+	public static final ConfigOption<Boolean> MATCH_CONTAINER_VCORES =
+			key("$internal.yarn.resourcemanager.enable-vcore-matching")
+					.booleanType()
+					.defaultValue(false)
+					.withDescription("**DO NOT USE** Whether YarnResourceManager should match the container vcores.");
```
Does this mean that one has to configure ones Flink cluster depending on the configuration of the Yarn cluster? What happens if one forgets about Flink?

xintongsong replied:
For the time being, yes.

Hadoop supports programmatically get this configuration from `RegisterApplicationMasterResponse` starting from 2.6.x, so we don't need users to manually configure this. But I believe the lowest Hadoop version Flink supports is 2.4.x, so we do not have a good way other than having user configure it.

This option should only affect the cases with dynamic worker resources. If the option is not set on a Yarn cluster that matches vcores, then workers with different cpu but same memory may not be schedules correctly. E.g., if Flink wants to start task executor `t1` in a container with resources `<1GB, 1 vcore>`, and `t2` in a container with resources `<1GB, 2 vcore>`, the actually resource available to `t1` might be `<1GB, 2 vcore>` and to `t2` might be `<1GB, 1 vcore>`.

tillrohrmann replied:
Ok, thanks for the clarification. Automatically configuring this flag with Hadoop >= 2.6 sounds like a good idea.

tillrohrmann commented on flink-yarn/src/test/java/org/apache/flink/yarn/YarnResourceManagerTest.java:
```diff
@@ -584,6 +590,85 @@ public void testGetCpuExceedMaxInt() throws Exception {
+		final int minVcore = 10;
+		final YarnResourceManager.WorkerSpecContainerResourceAdapter adapter =
+			new YarnResourceManager.WorkerSpecContainerResourceAdapter(
+				getConfigProcessSpecEqualsWorkerSpec(), minMemMB, minVcore, true);
```
It is usually easier to understand if one use an enum instead of boolean because one can give the different values expressive names (e.g. `MATCH_VCORES`, `IGNORE_VCORES`).

tillrohrmann commented on flink-yarn/src/test/java/org/apache/flink/yarn/TestingContainer.java:
```diff
@@ -0,0 +1,113 @@
+/**
+ * A {@link Container} implementation for testing.
+ */
+class TestingContainer extends Container {
```
I like the idea of getting rid of Mockito but how do we ensure that this works with all Hadoop versions? Looking at Hadoop 2.10. https://github.com/apache/hadoop/blob/release-2.10.0-RC1/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-api/src/main/java/org/apache/hadoop/yarn/api/records/Container.java it looks as if the container has gotten some more methods.

xintongsong replied:
Thanks for the pointer. Then I think we can extends `ContainerPBImpl` instead, similar to `TestingYarnAMRMClientAsync`.

I think what we really need is to override the following methods.
- `getId`
- `getNodeId`
- `getResource`
- `getPriority`

I'm aware that `ContainerPBImpl` is annotated `Unstable`, but we need these four methods anyway. If a future Hadoop version removed either of the four methods, then the original `mockContainer` will not work either.

tillrohrmann replied:
Using `ContainerPBImpl` should work.

tillrohrmann commented on flink-yarn/src/test/java/org/apache/flink/yarn/TestingContainerStatus.java:
```diff
@@ -0,0 +1,86 @@
+
+	@Override
+	public void setContainerId(ContainerId containerId) {
+
```
Shouldn't we fail in case someone calls `setContainerId` here? Otherwise it might go unnoticed and result in some strange behaviour/failure.

tillrohrmann commented on flink-yarn/src/test/java/org/apache/flink/yarn/TestingYarnAMRMClientAsync.java:
```diff
@@ -0,0 +1,113 @@
+/**
+ * A Yarn {@link AMRMClientAsync} implementation for testing.
+ */
+public class TestingYarnAMRMClientAsync extends AMRMClientAsyncImpl<AMRMClient.ContainerRequest> {
```
`AMRMClientAsyncImpl` is annotated as unstable. I'm not sure how will the testing implementation works across different Yarn versions. Have we tried this out?

xintongsong replied:
Same here.

I'm aware `AMRMClientAsyncImpl` is annotated as `Unstable`. But we are only overriding the minimum set of public methods here. If these methods are changed, then neither the original testing codes nor the production codes of Flink could work anymore.

I have not tried all the Yarn versions. But this should at least work with hadoop 2.4.1 (the default hadoop version if build from sources locally) and 2.8.3 (hadoop version for travis ci tests). I can further verify it with other versions.

tillrohrmann replied:
I think my problem with this class is that we are partially overriding an implementation class here. We don't really know how the overridden methods are used internally. Looking at the code, it looks as if they are all forwarding the calls to `client`. From this perspective it might look fine but if you take a look at `getAvailableResources`, which has not been overridden, then it also calls `client`. Do we know that there is no contract between the overridden methods and `getAvailableResources`? What if the `client` returns different available resources depending on how many container requests have been added? I'm not saying that this is the case, but I want to make that overriding individual methods of an implementation class can cause failures which are hard to predict and even harder to debug.

xintongsong replied:
I think we cannot guarantee that there's no contract between the overridden methods and the not-overridden methods. Even we can, these internal things might change in future.

On the other hand, what we can guarantee is that all the methods used by Flink are overridden. It won't be a problem if an un-overridden method is internally related to the overridden methods but is never called.

To that end, I guess we need to make the following changes.
- Override all the methods used in Flink. That means `setHeartbeatInterval` and `unregisterApplicationMaster` in addition to those already overridden.
- Explicitly state in the `javadoc` that all methods used in Flink should be overridden. That might help debug the tests if later Flink uses more public API of `AMRMClientAsync`.

tillrohrmann replied:
I think this is the price we have to pay for integrating with another system. It is always painful and tedious to set up the tests. In parts, this always involves reimplementing the other systems behaviour in mocks. I don't see a better atm. Hence, +1 for what you've proposed.

tillrohrmann commented on flink-yarn/src/test/java/org/apache/flink/yarn/TestingYarnAMRMClientAsync.java:
```diff
@@ -0,0 +1,113 @@
+
+	private Function<Tuple4<Priority, String, Resource, CallbackHandler>, List<? extends Collection<AMRMClient.ContainerRequest>>>
+		getMatchingRequestsFunction = ignored -> Collections.emptyList();
+	private Consumer<Tuple2<AMRMClient.ContainerRequest, CallbackHandler>> addContainerRequestConsumer = ignored -> {};
+	private Consumer<Tuple2<AMRMClient.ContainerRequest, CallbackHandler>> removeContainerRequestConsumer = ignored -> {};
+	private Consumer<Tuple2<ContainerId, CallbackHandler>> releaseAssignedContainerConsumer = ignored -> {};
```
I'd suggest to use the `BiConsumer` here.

tillrohrmann commented on flink-yarn/src/test/java/org/apache/flink/yarn/TestingYarnNMClientAsync.java:
```diff
@@ -0,0 +1,82 @@
+/**
+ * A Yarn {@link NMClientAsync} implementation for testing.
+ */
+class TestingYarnNMClientAsync extends NMClientAsyncImpl {
```
Same here `NMClientAsyncImpl` seems to be unstable and might change depending on the used Yarn version.

tillrohrmann commented on flink-yarn/src/test/java/org/apache/flink/yarn/TestingYarnNMClientAsync.java:
```diff
@@ -0,0 +1,82 @@
+ */
+class TestingYarnNMClientAsync extends NMClientAsyncImpl {
+
+	private Consumer<Tuple3<Container, ContainerLaunchContext, CallbackHandler>> startContainerAsyncConsumer = ignored -> {};
+	private Consumer<Tuple3<ContainerId, NodeId, CallbackHandler>> stopContainerAsyncConsumer = ignored -> {};
```
I'd suggest to use the `TriConsumer` here.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -363,31 +356,64 @@ public void onContainersCompleted(final List<ContainerStatus> statuses) {
-
-			excessContainers.forEach(this::returnExcessContainer);
-			requiredContainers.forEach(this::startTaskExecutorInContainer);
+			groupContainerByResource(containers).forEach(this::onContainersOfResourceAllocated);
```
I'd suggest to use the for-each loop.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -363,31 +356,64 @@ public void onContainersCompleted(final List<ContainerStatus> statuses) {
+
+	private void onContainersOfResourceAllocated(Resource resource, List<Container> containers) {
+		final List<WorkerResourceSpec> pendingWorkerResourceSpecs =
+			workerSpecContainerResourceAdapter.getWorkerSpecs(resource).stream()
```
Can it happen that `getWorkerSpecs(resource)` returns a list which contains a `WorkerResourceSpec` twice?

xintongsong replied:
It shouldn't. I'll change the return type of `getWorkerSpecs` to `Set`.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -412,30 +439,32 @@ private void releaseFailedContainerAndRequestNewContainerIfRequired(ContainerId
-
+	private void removeContainerRequest(AMRMClient.ContainerRequest pendingContainerRequest, WorkerResourceSpec workerResourceSpec) {
+		log.info("Removing container request {}.", pendingContainerRequest);
+		pendingWorkerCounter.decreaseAndGet(workerResourceSpec);
```
What about removing it from `workerSpecContainerResourceAdapter`?

tillrohrmann replied:
Differently asked, when do we clean `workerSpecContainerResourceAdapter` up?

xintongsong replied:
Unfortunately, we don't clean `workerSpecContainerResourceAdapter` up.

I think the upper bond of the amount of records really depends on how many different `WorkerResourceSpec` do we have. If we want to clean the unused records up, the `WorkerSpecContainerResourceAdapter` will need `YarnResourceManager` to tell it which WorkerResourceSpec is no longer needed (all corresponding TMs are completed and no pending ones). I'm not sure whether this complexity is necessary.

tillrohrmann replied:
Hmm, would it be possible to have a periodic cleanup task?

tillrohrmann replied:
Not saying that we have to implement it right away. I just want to know how one could fix this as a follow-up task.

xintongsong replied:
True, a periodic cleanup task should work there, or we can trigger a clean-up if the number of entries stored in `WorkerSpecContainerResourceAdapter` reaches a certain threshold.

I think the clean-up is definitely doable. It's just not very necessary at the time being.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -540,39 +571,41 @@ private FinalApplicationStatus getYarnStatus(ApplicationStatus status) {
+			.filter(entry ->
+				getContainerResource(entry.getKey()).equals(containerResource) &&
+				entry.getValue() > pendingWorkerCounter.getNum(entry.getKey()))
+			.findAny()
```
Shouldn't we do this for all instead of any?

xintongsong replied:
I think this is the same issue as `KubernetesResourceManager#requestKubernetesPodIfRequired`. I was trying to keep the exactly same behavior as before. But I think you're right, it is nice to request all required pods early.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -540,39 +571,41 @@ private FinalApplicationStatus getYarnStatus(ApplicationStatus status) {
-			requestYarnContainer();
-		}
+	private void requestYarnContainerIfRequired(Resource containerResource) {
+		getPendingWorkerNums().entrySet().stream()
```
I have to admit that I find `getPendingWorkerNums()` and `pendingWorkerCounter` quite confusing. The sound almost the same but the former means the requirements of the `SlotManager` and the latter the currently pending workers which have been requested by the RM.

xintongsong replied:
What's your suggestion here? How about renaming `getPendingWorkerNums` to `getSlotManagerPendingWorkerNums`?

tillrohrmann replied:
maybe we could rename `getPendingWorkerNums` into `getRequiredWorkers` or `getRequiredResources`.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -540,39 +571,41 @@ private FinalApplicationStatus getYarnStatus(ApplicationStatus status) {
-		if (requiredTaskManagers > numPendingContainerRequests) {
-			requestYarnContainer();
-		}
+	private void requestYarnContainerIfRequired(Resource containerResource) {
```
Shouldn't we iterate over all resources which are needed instead of restricting it to `containerResource`?

tillrohrmann commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/entrypoint/KubernetesResourceManagerFactoryTest.java:
```diff
@@ -0,0 +1,65 @@
+/**
+ * Tests for {@link KubernetesResourceManagerFactory}.
+ */
+public class KubernetesResourceManagerFactoryTest {
```
```suggestion
public class KubernetesResourceManagerFactoryTest extends TestLogger {
```

tillrohrmann commented on flink-yarn/src/test/java/org/apache/flink/yarn/entrypoint/YarnResourceManagerFactoryTest.java:
```diff
@@ -0,0 +1,79 @@
+/**
+ * Tests for {@link YarnResourceManagerFactory}.
+ */
+public class YarnResourceManagerFactoryTest {
```
```suggestion
public class YarnResourceManagerFactoryTest extends TestLogger {
```

xintongsong commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -232,57 +229,75 @@ private void recoverWorkerNodesFromPreviousAttempts() throws ResourceManagerExce
+	}
+
+	private KubernetesTaskManagerParameters createKubernetesTaskManagerParameters(WorkerResourceSpec workerResourceSpec) {
+		// TODO: need to unset process/flink memory size from configuration if dynamic worker resource is activated
```
The purpose of this `TODO` is to avoid conflict between `workerResourceSpec` and `process.size`/`flink.size` in `flinkConfig`. It is not about forbidden workers with different sizes. Actually, one of the main purpose of this PR is to make the RMs not assuming workers have the same size.

I was thinking about change the `Configuration` at creating the `SlotManager`, where we known which implementation of `SlotManager` is used and unset flink/process size if the plugin supports dynamic worker sizes.

tillrohrmann replied:
So do we have the requirement that what's written in the config and what's specified in `workerResourceSpec` have to be the same? If yes, then I think we should add a check state here. This will ensure that we remember what needs to be adjusted in order to support dynamic worker resources.

xintongsong replied:
I don't think we have that requirement.

The purpose of unsetting flink/process size from configuration, is to make sure the values in `workerResourceSpec` (task heap, task off-heap, network, managed) and those not in `workerResorceSpec`(framework heap, framework off-heap, jvm metaspace, jvm overhead, total flink, total process) can put together w/o conflict.

If they cannot be put together, we don't really need a check state to remind us because the generating of `taskExecutorProcessSpec` will fail.

tillrohrmann replied:
Ok, instead of creating a TODO, I would suggest to create a JIRA ticket which is linked as a follow up. TODO's tend to be forgotten.

xintongsong replied:
I've created FLINK-17061 to track this issue.

xintongsong commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesResourceManagerTest.java:
```diff
@@ -321,6 +339,47 @@ public void testGetCpuCoresNumSlots() {
+
+		// Terminate pod1.
+		terminatePod(initialPod1);
+		resourceManager.setCustomPendingWorkerNums(Collections.singletonMap(workerResourceSpec1, 1));
```
OK, I'll try to pass in a testing `SlotManager` and control what to be returned from `getPendingWorkerNums`. How does that sound like?

tillrohrmann replied:
This sounds good.

xintongsong commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -615,4 +632,85 @@ protected double getCpuCores(final Configuration configuration) {
+		}
+
+		private Resource createAndMapContainerResource(final WorkerResourceSpec workerResourceSpec) {
+			// TODO: need to unset process/flink memory size from configuration if dynamic worker resource is activated
```
Same here. I think the RM implementations should not be aware of whether the dynamic worker resources is activated or not.

tillrohrmann replied:
In general I agree unless it is not yet fully supported by the RM implementation. If it is, then I guess we also don't need this comment. If it is not supported, then the check state will help us to remember what needs to be changed.

xintongsong replied:
I think after these PRs, the RM implementations should fully support dynamic worker resources. The problem is that we not yet have a SM implementation that supports it. Once we have a SM implementation that supports dynamic worker resources, we should unset the process/flink size from configuration when using it. And if we don't do the unset, Flink might fail at the places of this `TODO`. So I added this `TODO` as a reminder, that if later these codes fail this might be the reason.

tillrohrmann replied:
The current SM will be initialized with the proper default `WorkerResourceSpec` derived from Flink's configuration, right? Hence, we could already resolve this TODO, right? Then there would not be any problems with enabling support for dynamic worker resources.

xintongsong commented:
@tillrohrmann, thanks for the review.

Regarding hiding `pendingWorkerCounter` behind `ActiveResourceManager` methods, I think this is indeed a good suggestion. Just trying to confirm which methods the base class should expose before working on the implementation. I think the following methods would be good enough. WDYT?
- `getNumPendingWorkers()`
- `getNumPendingWorkers(WorkerResourceSpec workerResourceSpec)`
- `notifyNewWorkerRequested(WorkerResourceSpec workerResourceSpec)`
- `notifyNewWorkerAllocated(WorkerResourceSpec workerResourceSpec)`

BTW, I was trying to split the changes for K8s/Yarn into 2 PRs (#11323 for K8s and this one for Yarn) to make the review easier. I mentioned in the PR description that this PR is based on #11323, but I guess it's not obvious enough. So, would you prefer that I address the comments related to k8s in #11323, or just keep them together in this PR? I'm ok with both ways.

tillrohrmann commented:
I would name the methods the following:

* `getNumPendingWorkers()`
* `getNumPendingWorkersFor(WorkerResourceSpec workerResourceSpec)`
* `notifyNewWorkerRequested(WRS wrs)`
* `notifyNewWorkerAllocated(WRS wrs)`

I actually think that neither the Yarn nor the K8s `RM` implementation should need to call these methods if we did the abstraction correctly. This is of course unrelated to these PRs. I believe that we could move a lot of functionality from the specific implementation into the parent class.

I don't have a strong preference concerning how to split the changes. Do what's easiest for you.

xintongsong commented:
@tillrohrmann 
I have addressed the comments. Please take a look.

This PR has been rebased onto recent master, for conflict resolving.

A hotfix commit has been added to fix that `PerJobMiniClusterFactory` improperly calculate `numSlotsPerTaskManager`, introduced by FLINK-16705.

tillrohrmann commented on flink-yarn/src/test/java/org/apache/flink/yarn/RegisterApplicationMasterResponseReflectorTest.java:
```diff
@@ -88,7 +94,44 @@ public void testGetMethodReflectiveHadoop22() {
+			new RegisterApplicationMasterResponseReflector(LOG, HasMethod.class);
+
+		final Optional<Set<String>> schedulerResourceTypeNames =
+			registerApplicationMasterResponseReflector.getSchedulerResourceTypeNamesUnsafe(new HasMethod());
```
Instead of calling private methods of the `RegisterApplicationMasterResponseReflector` we could also add an `assumeTrue` statement based on the Hadoop version. Then we can have two tests for Hadoop >= 2.6 and < 2.6.

xintongsong replied:
I'm not sure about this. My concern is that, the <2.6 test case might eventually not executed in most cases.

Take `testDoesntCallGetContainersFromPreviousAttemptsMethodIfAbsent` as an example. If we make this test case only executed with Hadoop <2.2, then it's practically not executed. Currently we have Hadoop 2.8 for travis ci test, 2.4 & 2.8 for nightly test, and 2.4 (pom default) for local maven verify unless another version is intentionally specified.

tillrohrmann replied:
Makes sense. I agree that your proposal is better.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -117,7 +118,9 @@
 	private NMClientAsync nodeManagerClient;
 
-	private final WorkerSpecContainerResourceAdapter workerSpecContainerResourceAdapter;
+	private WorkerSpecContainerResourceAdapter workerSpecContainerResourceAdapter = null;
```
An idea to keep the `WorkerSpecContainerResourceAdapter` final is to pass in the `WorkerSpecContainerResourceAdapter.MatchingStrategy` to the method `getWorkerSpecs`. If you look at the `WorkerSpecContainerResourceAdapter` class then one also sees that the matching strategy is not really an essential part of it. Only the lookup method changes its behaviour based on it. Everything else stays the same.

xintongsong commented on flink-yarn/src/test/java/org/apache/flink/yarn/YarnResourceManagerTest.java:
```diff
@@ -243,24 +248,33 @@ protected NMClientAsync createAndStartNodeManagerClient(YarnConfiguration yarnCo
-				.build();
+		Context(Configuration configuration, @Nullable SlotManager slotManager) throws  Exception {
+
+			workerResourceSpec = YarnWorkerResourceSpecFactory.INSTANCE.createDefaultWorkerResourceSpec(configuration);
+			if (slotManager == null) {
+				slotManager = SlotManagerBuilder.newBuilder()
+					.setDefaultWorkerResourceSpec(workerResourceSpec)
+					.build();
+			}
```
It requires `workerResourceSpec` for creating the default `slotManager`.
I was trying to avoid creating `workerResourceSpec` twice, in both constructors.

tillrohrmann replied:
I think it would be fine to add a `SlotManager.getDefaultWorkerResourceSpec()`.

xintongsong commented on flink-yarn/src/test/java/org/apache/flink/yarn/TestingYarnAMRMClientAsync.java:
```diff
@@ -0,0 +1,158 @@
+		unregisterApplicationMasterConsumer.accept(appStatus, appMessage, appTrackingUrl);
+	}
+
+	void setGetMatchingRequestsFunction(
+		Function<Tuple4<Priority, String, Resource, CallbackHandler>, List<? extends Collection<AMRMClient.ContainerRequest>>>
+			getMatchingRequestsFunction) {
+		this.getMatchingRequestsFunction = Preconditions.checkNotNull(getMatchingRequestsFunction);
+	}
+
+	void setAddContainerRequestConsumer(
+		BiConsumer<AMRMClient.ContainerRequest, CallbackHandler> addContainerRequestConsumer) {
+		this.addContainerRequestConsumer = Preconditions.checkNotNull(addContainerRequestConsumer);
+	}
+
+	void setRemoveContainerRequestConsumer(
+		BiConsumer<AMRMClient.ContainerRequest, CallbackHandler> removeContainerRequestConsumer) {
+		this.removeContainerRequestConsumer = Preconditions.checkNotNull(removeContainerRequestConsumer);
+	}
+
+	void setReleaseAssignedContainerConsumer(
+		BiConsumer<ContainerId, CallbackHandler> releaseAssignedContainerConsumer) {
+		this.releaseAssignedContainerConsumer = Preconditions.checkNotNull(releaseAssignedContainerConsumer);
+	}
+
+	void setSetHeartbeatIntervalConsumer(
+		Consumer<Integer> setHeartbeatIntervalConsumer) {
+		this.setHeartbeatIntervalConsumer = setHeartbeatIntervalConsumer;
+	}
+
+	void setRegisterApplicationMasterFunction(
+		TriFunction<String, Integer, String, RegisterApplicationMasterResponse> registerApplicationMasterFunction) {
+		this.registerApplicationMasterFunction = registerApplicationMasterFunction;
+	}
+
+	void setUnregisterApplicationMasterConsumer(
+		TriConsumer<FinalApplicationStatus, String, String> unregisterApplicationMasterConsumer) {
+		this.unregisterApplicationMasterConsumer = unregisterApplicationMasterConsumer;
+	}
```
The reason I did not set these functions in the constructor is that, one would have to construct the entire `TestingYarnAMRMClientAsync` for defining one of these functions in a test case. We have several such testing classes (RM client, NM client, slot manager), and it would be complicated if some of the test cases need to provide customized instances for a subset of these classes when creating the `Context`, while some other test cases need to do that for a different subset.

I think making the fields `volatile` sounds good to me.

tillrohrmann replied:
Makes sense.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/RegisterApplicationMasterResponseReflector.java:
```diff
@@ -96,7 +110,40 @@
+	}
+
+	@VisibleForTesting
+	Optional<Set<String>> getSchedulerResourceTypeNamesUnsafe(final Object response) {
```
Won't do because of testing purposes.

tillrohrmann commented:
Rebased onto the latest state of #11323.
